### PR TITLE
Generalize factory and device config commands, add ability to manage device group config

### PR DESCRIPTION
**Pull request by vkhoroz**
There is still work to do with a non-trivial updates subcommand, but it's already possible to review config log/set/delete subcommands, all of which now behave consistently for factory/group/device config.

### CODE REVIEW

File: client/foundries.go [hunk 1]
```diff
@@ -682,10 +682,53 @@ func (a *Api) FactoryPatchConfig(factory string, cfg ConfigCreateRequest) error
...
+	return a.DeviceListConfigCont(url)
+}
+
+func (a *Api) GroupCreateConfig(factory, group string, cfg ConfigCreateRequest) error {
+	data, err := json.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+
+	url := a.serverUrl + "/ota/factories/" + factory + "/device-groups/" + group + "/config/"
```
Comment (mike-sul) [inline client/foundries.go:699]: So, a config filename is specified in the POST body, not in URL?

File: client/foundries.go [hunk 2]
```diff
@@ -682,10 +682,53 @@ func (a *Api) FactoryPatchConfig(factory string, cfg ConfigCreateRequest) error
...
+}
+
+func (a *Api) GroupPatchConfig(factory, group string, cfg ConfigCreateRequest) error {
+	data, err := json.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+
+	url := a.serverUrl + "/ota/factories/" + factory + "/device-groups/" + group + "/config/"
+	logrus.Debug("Creating new device group config")
```
Comment (mike-sul) [inline client/foundries.go:719]: `Patching` instead of `Creating`?

File: client/foundries.go [hunk 3]
```diff
@@ -682,10 +682,53 @@ func (a *Api) FactoryPatchConfig(factory string, cfg ConfigCreateRequest) error
...
+	return err
+}
+
+func (a *Api) GroupPatchConfig(factory, group string, cfg ConfigCreateRequest) error {
+	data, err := json.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+
+	url := a.serverUrl + "/ota/factories/" + factory + "/device-groups/" + group + "/config/"
```
Comment (mike-sul) [inline client/foundries.go:718]: Don't we need a filename of a config group that is being patched?

**Comment by mike-sul** [comment]
I hardly understand the current API of device and factory configuration management, so it makes me an incompetent reviewer :).

File: client/foundries.go (hunk 1 above)
Comment (vkhoroz) [reply client/foundries.go:702]: Yes

Comment (vkhoroz) [reply client/foundries.go:702]: and there could be multiple filenames at once

File: client/foundries.go (hunk 3 above)
Comment (vkhoroz) [reply client/foundries.go:721]: A patch can apply to entire config which has many filenames

File: client/foundries.go (hunk 2 above)
Comment (vkhoroz) [reply client/foundries.go:725]: I did the same as it is for factory/device configs

**Review by doanac — COMMENTED** [review]
looks good. just a few minor things.

**Comment by mike-sul** [comment]
How to list devices of a specific device group?

File: subcommands/config/log.go [hunk 4]
```diff
@@ -12,32 +12,36 @@ import (
 func init() {
 	logCmd := &cobra.Command{
 		Use:   "log",
-		Short: "Changelog of configuration",
+		Short: "Show a change log of configuration",
```
[the remark this answers is not in the document]
Comment (vkhoroz) [reply subcommands/config/log.go:15]: ok

File: subcommands/common_config.go [hunk 5]
```diff
@@ -42,6 +80,21 @@ func LogConfigs(opts *LogConfigsOptions) {
 	}
 }
 
+func ReadConfig(configFile string, cfg *client.ConfigCreateRequest) {
```
[the remark this answers is not in the document]
Comment (vkhoroz) [reply subcommands/common_config.go:96]: To me it's a mirror of `PrintConfig` which is already used outside.
A `ReadConfig` is only only used inside this module ATM, but I don't consider it to be private.

File: subcommands/common_config.go [hunk 6]
```diff
@@ -127,3 +137,146 @@ func PrintConfig(cfg *client.DeviceConfig, showAppliedAt, highlightFirstLine boo
 		}
 	}
 }
+
+type SetUpdatesConfigOptions struct {
+	UpdateTags     string
```
[the remark this answers is not in the document]
Comment (vkhoroz) [reply subcommands/common_config.go:142]: ok

**Comment by vkhoroz** [comment]
> 
> 
> How to list devices of a specific device group?

This was already implemented in #67 .
`fioctl devices list -f <factory> -g <group>` or just a group if a factory goes from your config file.

File: subcommands/common_config.go [hunk 7]
```diff
@@ -139,7 +139,7 @@ func PrintConfig(cfg *client.DeviceConfig, showAppliedAt, highlightFirstLine boo
 }
 
 type SetUpdatesConfigOptions struct {
-	UpdateTags     string
+	UpdateTag      string
```
Comment (doanac) [inline subcommands/common_config.go:142]: sorry - i wasn't clear in my previous comment. this is kind of backwards. I was trying to say - make the CLI only accept "updateTag". The backend is still technically "tags". But lets change fioctl to only let people configure a single tag. 

If this is a pain, we can drop this commit and do it as a separate PR?

Comment (vkhoroz) [reply subcommands/common_config.go:142]: Then I prefer to drop this commit.

A change you propose would be backward incompatible and might break some users expectations.
I'm not a fan to add a new logic and break an old logic within one feature.

It's not so trivial to be able to preserve multi-tag behavior for already present configs, and only disallow it for new configs.

**Review by doanac — APPROVED** [review]
